prune: Parallelize repack step

MichaelEischer opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------
The PR parallelizes the repack step of the prune command.

This change is basically a simpler reimplementation of the corresponding part of #2340. There are a few notable differences though:
- The repack pipeline in the PR only has 3 steps instead of 4 in #2340. The steps are: send packIDs into the pipeline, download packs, process pack + store blobs. In #2340 storing blobs is a separate pipeline step. However, this requires an additional buffer pool. Parallelizing the actual upload should be handled via #2696.
- Repacked blobs are still removed from `keepBlobs` and not tracked in an additional sync.Map. The implementation follows #2340 in that it  is prepared to support using one of multiple duplicates of a blob in the future. (This is the reason for the duplicate keepBlobs check)
- The number of workers is currently fixed at 8, adapting this depending on the backend and CPU core available should happen in another PR which adjusts other parts of restic as well.
- Usage of sync.errgroup instead of tomb.

Note: Most of the changes are indentation only, so the actual changes are best viewed when ignoring whitespace changes.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------
Probably in one of the issues regarding prune performance.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- ~~[ ] I have added tests for all changes in this PR~~ **Repack already has tests**
- ~~[ ] I have added documentation for the changes (in the manual)~~
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

MichaelEischer committed:
prune: Parallelize repack command

J0WI commented:
IMHO it would make sense to review and merge https://github.com/restic/restic/pull/2844 first before changing prune.

MichaelEischer commented:
> IMHO it would make sense to review and merge #2844 first before changing prune.

That's the plan.

I've primarily opened the PR to avoid duplicate work. (Well, actually it was quicker to code this and run prune afterwards than wait for the four hour unparallelized repack step to complete ;-) )

greatroar commented:
I only took a cursory glance at the code, but could you explain why there are still two worker pools connected by a channel? It looks like the downloader goroutines could just call the processing code synchronously. Then numRepackWorkers would be the actual number of workers, rather than half of that number.

MichaelEischer commented:
@greatroar The idea is to split the workers in the future into CPU- and IO-bound (to some extent). That is the number of download workers would scale with number of connections the backend can handle, whereas the repacking itself is more CPU-bound and would be limited to `GOMAXPROCS` (more workers would just unnecessarily gobble up RAM, but provide no speed-up). Right now the second part of the pipeline also handles the uploads itself and thus alternates between CPU/IO-bound.

I'm aware that `DownloadAndHash` will also require a lot CPU to verify the pack hash, but I want to keep the check of the downloaded pack file near the download itself to e.g. retry the download of damaged packs in the future. Repacking blobs is probably more expensive as it has to decrypt and hash each blob, then encrypt the blob and has the new pack file.

Right now, the number of repack workers is just an arbitrary number (I just used the same value as in `cmd/restic/delete.go`).

fd0 committed:
Clarify message about expected error

fd0 committed:
Use a non-constant seed

fd0 committed:
Fix typo

fd0 committed:
Remove unneeded seek

The file returned from DownloadAndHash() is already seeked to the start
of the file.

fd0 commented on internal/repository/repack.go:
```diff
@@ -22,91 +27,166 @@ func Repack(ctx context.Context, repo restic.Repository, packs restic.IDSet, kee
+						h, tempfile.Name(), id)
+				}
+
+				keepMutex.Lock()
+				// recheck whether some other worker was faster
+				shouldKeep = keepBlobs.Has(h)
+				if shouldKeep {
+					keepBlobs.Delete(h)
+				}
+				keepMutex.Unlock()
```
This modifies the map passed from the caller (maps are call-by-reference). How about we use a new (local) map to keep track of which blobs were already processed?

MichaelEischer replied:
I'd actually prefer to stick to the current way of modifying the map: the list of blobs to keep can be one the main contributors to the memory usage of the prune command. When restic has to rewrite a large fraction of the packs in a repository, then your suggested `rewrittenBlobs` map would in the worst case contain nearly all existing blobs. That could increase the memory usage of prune by up to 50% (compared to restic with merged #2718 and #2842. In that constellation the repository index and the usedBlobs map should consume roughly similar amounts of memory).

fd0 replied:
Fine with me, I'll add a note to the function's documentation then we're good to go.

fd0 reviewed (changes requested):
Great work! I'd love to refactor this into smaller functions (and maybe a `Repacker` struct which bundles all things we need nicely), but this will do for now.

We just need to clarify what we're going to do with the map, then it's ready to be merged.

fd0 committed:
Add comment that keepBlobs is modified
